api/utils/TokenChecker.py

```diff
@@ -0,0 +1,194 @@
+"""This module contains the TokenChecker class."""
+import base64
+import datetime
+import httplib
+import json
+import logging
+import urllib2
+
+from django.conf import settings
+from django.core.cache import cache
+from jose import jwt
+from jose.exceptions import ExpiredSignatureError, JWTClaimsError, JWTError
+
+
+class TokenChecker:
+    """This class contains methods to check a JWT token for validity."""
+
+    def __init__(self):
+        """Initialize a new TokenChecker."""
+        self.logger = logging.getLogger(__name__)
+
+    def valid_token_to_id(self, token):
+        """Introspect a token to determine it's origin."""
+        try:
+            jwt_unverified_json = jwt.get_unverified_claims(token)
+        except JWTError:
+            self.logger.error('Token cannot be decoded.')
+            self.logger.debug('Full token: %s', token)
+            return None
+
+        unverified_token_id = jwt_unverified_json['sub']
+        self.logger.info('Token claims to be from %s', unverified_token_id)
+
+        # if token is in the cache, we say it is valid
+        if cache.get(unverified_token_id) == token:
+            self.logger.info("Token is in cache.")
+
+            return jwt_unverified_json['sub']
+
+        # otherwise, we need to validate the token
+        if not self._is_token_json_temporally_valid(jwt_unverified_json):
```

Comment:
checking whether it's in the cache and whether it's valid? Not really... but I don't think there's a reason to check the validity of a token that is in the cache because only valid tokens get put in the cache, so the only reason a previously valid token could become invalid is it is revoked or it expires. 

If it is revoked, it will take up to 5 minutes for the cached copies of the token to drop out of the cache and Indigo would prefer this to constantly contacting an IAM. If that's the policy for revoked cached tokens, it seems overkill (to me) to be more strict for expired tokens.

Reply:
Ok, was mostly just interested if there were any specific reasons. 😃